### ISSUE TITLE
How to use ApplyAdam(C++API)?

### ISSUE THREAD

**Issue by minjac** [tensorflow/tensorflow#37536]
Hi, I has used C++ API, but I got some problem about ApplyAdam:
  ApplyAdam(const ::tensorflow::Scope& scope, ::tensorflow::Input var,
          ::tensorflow::Input m, ::tensorflow::Input v, ::tensorflow::Input
          beta1_power, ::tensorflow::Input beta2_power, ::tensorflow::Input lr,
          ::tensorflow::Input beta1, ::tensorflow::Input beta2,
          ::tensorflow::Input epsilon, ::tensorflow::Input grad);

I do not know beta1_power and beta2_power in adam optimizer.

**Comment by Saduf2019** [comment]
@minjac 
Could you please share a simple standalone code to replicate the issue faced by you in our local,along with the tensor flow version.

**Comment by minjac** [comment]
> 
> 
> @minjac
> Could you please share a simple standalone code to replicate the issue faced by you in our local,along with the tensor flow version.

Hi, code is doesn't matter in this case, because I just don't know what the beta1_power and beta2_power mean. 

`/// Update '*var' according to the Adam algorithm.
///
/// $$lr_t := \text{learning\_rate} * \sqrt{1 - beta_2^t} / (1 - beta_1^t)$$
/// $$m_t := beta_1 * m_{t-1} + (1 - beta_1) * g$$
/// $$v_t := beta_2 * v_{t-1} + (1 - beta_2) * g * g$$
/// $$variable := variable - lr_t * m_t / (\sqrt{v_t} + \epsilon)$$
///
/// Arguments:
/// * scope: A Scope object
/// * var: Should be from a Variable().
/// * m: Should be from a Variable().
/// * v: Should be from a Variable().
/// * beta1_power: Must be a scalar.
/// * beta2_power: Must be a scalar.
/// * lr: Scaling factor. Must be a scalar.
/// * beta1: Momentum factor. Must be a scalar.
/// * beta2: Momentum factor. Must be a scalar.
/// * epsilon: Ridge term. Must be a scalar.
/// * grad: The gradient.`

As the comments says, I know all the arguments mean except the beta1_power and beta2_power, because I don't find where the beta1_power and beta2_power are in the Adam Optimizer's formula.

**Comment by Saduf2019** [comment]
@minjac 
can you please refer to [this link](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Adam?version=nightly) and let us know if it helps

**Comment by minjac** [comment]
> 
> 
> @minjac
> can you please refer to [this link](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Adam?version=nightly) and let us know if it helps
Thanks for your help, but I cannot find some arguments as same as beta1_power or beta2_power in this API.

**Comment by Saduf2019** [comment]
@minjac 
here is some info on the variables,with reference to your question "Hi, code doesn't matter in this case, because I just don't know what the beta1_power and beta2_power mean."

The beta-variables are variables used by the AdamOptimizer that need to be initialized similarily to your other variables. You can either initialize them with the tf.global_variables_initializer() after creating the optimizer, or look up the variables and initialize them directly with tf.variables_initializer().

here are few links related to them [link1](https://stackoverflow.com/questions/45498342/tensorflow-adamoptimizer-beta1-power-not-initialized) [link2](https://www.tensorflow.org/api_docs/python/tf/raw_ops/ResourceApplyAdamWithAmsgrad)

please let us know if this helps.

**Comment by Saduf2019** [comment]
@minjac
please update on the above comment

**Comment by minjac** [comment]
> 
> 
> @minjac
> please update on the above comment

Hi, thanks for reply. I click your link, **beta1_power** and **beta2_power** seems to be the same as **t** in Adam Optimizer's formula.

**Comment by ymodak** [comment]
From https://arxiv.org/pdf/1412.6980.pdf beta1, beta2 are exponential decay rates for the moment estimates.
(default values are set as β1 = 0.9, β2 = 0.999)
https://github.com/tensorflow/tensorflow/blob/e5bf8de410005de06a7ff5393fafdf832ef1d4ad/tensorflow/python/keras/optimizer_v2/adam.py#L48

**Comment by minjac** [comment]
> 
> 
> From https://arxiv.org/pdf/1412.6980.pdf beta1, beta2 are exponential decay rates for the moment estimates.
> (default values are set as β1 = 0.9, β2 = 0.999)
> 
> https://github.com/tensorflow/tensorflow/blob/e5bf8de410005de06a7ff5393fafdf832ef1d4ad/tensorflow/python/keras/optimizer_v2/adam.py#L51

Thanks for reply, beta_2 seems to be the same as beta1 which in my question.

**Comment by ymodak** [comment]
This question is better asked on [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) since it is not a bug or feature request. There is also a larger community that reads questions there.
Thanks!
